### PR TITLE
feat: Cypress Studio for Cypress 10

### LINKED ISSUE #23461: Prepare Studio Feature branch for review
Part of https://github.com/cypress-io/cypress/issues/23230

Studio support is in feature branch [feat/studio-cypress-10](https://github.com/cypress-io/cypress/tree/feat/studio-cypress-10)

Once all the other Studio stories are complete, we need to do a final pass on the feature branch, test it all out, and prepare it for a thorough review and QA. 

AC:

- it all works
- add lots of tests
- fix up any obvious tech debt from 9.x

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes #23461

### User facing changelog

Re-enable Cypress Studio for End to End testing via the `experimentalStudio` flag.

### Steps to test

Watch this video walkthrough on how the feature works (Cypress team only): https://cypressio.slack.com/archives/C011BAPC614/p1661517276740949?thread_ts=1661517030.718579&cid=C011BAPC614 (#team-component-testing channel)

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/4696
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/reporter/cypress/e2e/runnables.cy.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ describe('runnables', () => {
       cy.visit('/').then((win) => {
         win.render(Object.assign({
           runner,
+          studioEnabled: true,
```
Comment (emilyrohrbough) [inline packages/reporter/cypress/e2e/runnables.cy.ts:27]: We should only enable studio for a sub-set of thee tests since it's not always on

File: packages/reporter/cypress/e2e/suites.cy.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@ describe('suites', () => {
     cy.visit('/').then((win) => {
       win.render({
         runner,
+        studioEnabled: true,
```
Comment (emilyrohrbough) [inline packages/reporter/cypress/e2e/suites.cy.ts:18]: Same comment as above, this should only be enabled for a sub-set of tests

File: packages/reporter/cypress/e2e/commands.cy.ts [hunk 3]
```diff
@@ -958,7 +958,7 @@ describe('commands', { viewportHeight: 1000 }, () => {
   })
 
   // FIXME: When studio support is re-introduced we can enable these tests.
```
Comment (emilyrohrbough) [inline packages/reporter/cypress/e2e/commands.cy.ts:960]:
```suggestion

```

File: packages/reporter/src/commands/command.tsx [hunk 4]
```diff
@@ -471,6 +481,19 @@ class Command extends Component<Props> {
...
+
+  _removeStudioCommand = (e: React.MouseEvent<HTMLElement, globalThis.MouseEvent>) => {
+    e.preventDefault()
+    e.stopPropagation()
+
+    const { model, events } = this.props
+
+    if (!model.isStudio) {
+      return
+    }
```
Comment (emilyrohrbough) [inline packages/reporter/src/commands/command.tsx:493]: Don't think this is necessary since it only is applied to the studio-command-remove icon when model.isStudio.

File: packages/reporter/src/errors/errors.scss [hunk 5]
```diff
@@ -59,6 +59,14 @@ $code-border-radius: 4px;
     text-align: center;
   }
 
+  .studio-err-wrapper {
```
Comment (emilyrohrbough) [inline packages/reporter/src/errors/errors.scss:62]: Can we only render `.studio-err-wrapper` when an error occurs vs always?

File: packages/server/lib/socket-e2e.ts [hunk 6]
```diff
@@ -102,8 +102,7 @@ export class SocketE2E extends SocketBase {
         })
 
         socket.on('studio:init', (cb) => {
-          studio.getStudioModalShown()
-          .then(cb)
+          cb()
```
Comment (emilyrohrbough) [inline packages/server/lib/socket-e2e.ts:105]: If we aren't triggering anything, can we remove this event listener & remove the event emit from @packages/app?

File: packages/app/cypress/e2e/studio/README.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+## Cypress Studio Tests
+
+These are the tests for the Cypress Studio feature. [Learn more here](https://docs.cypress.io/guides/references/cypress-studio).
+
+Due to an (currently unknown) reason, if more than 1 test is written per spec file, the second test will fail. For this reason, each file has one test.
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/studio/README.md:5]: what is the second test's error message?

File: packages/app/cypress/e2e/studio/studio-kitchensink.cy.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
+import { launchStudio } from './helper'
+
+describe('Cypress Studio', () => {
+  it('kitchensink - writes a test with all kinds of assertions', () => {
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/studio/studio-kitchensink.cy.ts:4]: the name kitchensink make me thing we are testing the kitchen sink project.

File: packages/app/cypress/e2e/studio/studio-smoke.cy.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+      expect(spec.trim()).to.eq(`
+it('visits a basic html page', () => {
+  cy.visit('cypress/e2e/index.html')
+  /* ==== Generated with Cypress Studio ==== */
+  cy.get('#increment').click();
+  /* ==== End Cypress Studio ==== */
+})`.trim())
+    })
+  })
```
Comment (emilyrohrbough) [inline packages/app/cypress/e2e/studio/studio-smoke.cy.ts:41]: do we need to verify we can "run" the save spec successfully?

File: packages/config/src/options.ts [hunk 10]
```diff
@@ -223,6 +221,12 @@ const driverConfigOptions: Array<DriverConfigOption> = [
     validation: validate.isBoolean,
     isExperimental: true,
     requireRestartOnChange: 'server',
+  }, {
+    name: 'experimentalStudio',
+    defaultValue: false,
+    validation: validate.isBoolean,
+    isExperimental: true,
+    requireRestartOnChange: 'server',
```
Comment (emilyrohrbough) [inline packages/config/src/options.ts:229]: What is changing on the server side that requires a server restart?

File: packages/reporter/src/test/test.cy.tsx [hunk 11]
```diff
@@ -16,7 +16,11 @@ describe('test/test.tsx', () => {
...
     cy.mount(<div className="runnable suite">
-      <Test model={model} appState={appState} />
+      <Test
+        model={model}
+        appState={appState}
+        studioEnabled={false}
+      />
     </div>)
 
     cy.percySnapshot()
```
Comment (emilyrohrbough) [inline packages/reporter/src/test/test.cy.tsx:26]: Can we add a test with studioEnabled = true?

File: packages/reporter/src/errors/errors.scss (hunk 5 above)
Comment (lmiller1990) [reply packages/reporter/src/errors/errors.scss:62]: Yeah this was the Cypress 9 implementation, not really ideal... I'll see if I can quickly fix this.

File: packages/app/cypress/e2e/studio/studio-smoke.cy.ts (hunk 9 above)
Comment (lmiller1990) [reply packages/app/cypress/e2e/studio/studio-smoke.cy.ts:60]: Updated to assert the test passes and the correct commands are run.

https://github.com/cypress-io/cypress/pull/23544/files#diff-8c34fb9114d322c67ada04c39e99fc5ccdacf173ca98e3f901cd486919917ec8R42-R60

File: packages/config/src/options.ts (hunk 10 above)
Comment (lmiller1990) [reply packages/config/src/options.ts:229]: Nothing actually - good point. It looks like we do need to restart the browser though to re-initialize the `Reporter` with `config.experimentalStudio`'s new value, so I set this to `'browser'`.

File: packages/reporter/cypress/e2e/runnables.cy.ts (hunk 1 above)
Comment (lmiller1990) [reply packages/reporter/cypress/e2e/runnables.cy.ts:27]: Fixed

Also made this file more type safe and made the entire file more type safe... only to realize we don't seem to run `tsc` on this package on CI.

We should go through and properly type the entire of `reporter`.

File: packages/reporter/cypress/e2e/suites.cy.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/reporter/cypress/e2e/suites.cy.ts]: https://github.com/cypress-io/cypress/pull/23544/commits/5ffeae71dc69efaa0ee7697d13f3a81ea57bc02c

There is a lot of duplication in the `beforeEach` for almost every single spec in `Reporter` - I think we should go through, refactor a lot (if not all) to component tests and introduce some shared utils to reduce duplication.

Also made a small improvement in ^ to use the correct prop for the runnerStore... looks like there isn't any type checking in reporter, we definitely need to get more proactive with type safety.

File: packages/server/lib/socket-e2e.ts (hunk 6 above)
Comment (lmiller1990) [reply packages/server/lib/socket-e2e.ts]: done

File: packages/app/cypress/e2e/studio/README.md (hunk 7 above)
Comment (lmiller1990) [reply packages/app/cypress/e2e/studio/README.md]: No error, turns out it was a race condition and I fixed it. Now all the studios test are in a single `studio.spec.ts` file. :100:

Comment (lmiller1990) [reply packages/app/cypress/e2e/studio/README.md]: Commit consolidating them: https://github.com/cypress-io/cypress/pull/23544/commits/ffc10d415e1445af1c8b593f2736ad68276cd7fd

File: packages/reporter/src/test/test.tsx [hunk 12]
```diff
@@ -182,7 +183,7 @@ class Test extends Component<TestProps> {
   _controls () {
     let controls: Array<JSX.Element> = []
 
-    if (this.props.model.state) {
+    if (this.props.model.state === 'failed') {
```
Comment (astone123) [inline packages/reporter/src/test/test.tsx:186]: This is a drive-by fix. Right now we render this control even when the test hasn't failed

**Comment by ZachJW34** [comment]
I love how studio works with the SelectorPlayground, allowing you to have fine grained control on what you are selecting in the DOM!

File: packages/app/cypress/e2e/studio/studio.cy.ts [hunk 13]
```diff
@@ -0,0 +1,242 @@
+import { launchStudio } from './helper'
```
Comment (ZachJW34) [inline packages/app/cypress/e2e/studio/studio.cy.ts:1]: These test are failing when I run them locally, but are [passing in CI](https://app.circleci.com/pipelines/github/cypress-io/cypress/42523/workflows/a852ef89-7559-4da2-8bf6-738ac8949d51/jobs/1768275/parallel-runs/7?filterBy=ALL)? Is there something I'm missing when running these tests?

Screen Shot 2022-08-26 at 9 55 31 AM

**Review by ZachJW34 — APPROVED** [review]
Nothing blocking from my comments, mostly nits (and I'm curious about the tests). Going to test on my windows real quick.

Works great!!

File: packages/app/src/runner/studio/StudioInstructionsModal.vue [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    :help-text="t('links.learnMoreButton')"
+    variant="bare"
+    data-cy="studio-instructions-modal"
+    @update:model-value="emit('close')"
+  >
+    <template #title>
+      {{ t('runner.studio.studio') }} {{ t('versions.beta') }}
+    </template>
+
+    <div class="py-7 px-15 text-gray-900">
```
[the remark this answers is not in the document]
Comment (astone123) [reply packages/app/src/runner/studio/StudioInstructionsModal.vue]: Just pushed a commit up that fixes this 👍🏻

File: packages/frontend-shared/src/components/StandardModal.vue [hunk 15]
```diff
@@ -17,6 +17,7 @@
         :class="props.class || ''"
       >
         <StandardModalHeader
+          :no-help="noHelp"
```
[the remark this answers is not in the document]
Comment (astone123) [reply packages/frontend-shared/src/components/StandardModal.vue:20]: Yeah that's the way I had it before but Lachlan updated it to have this prop for some reason, I think it had to do with testing? I think we'll leave it for now

File: packages/app/src/runner/studio/StudioControls.vue [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+          v-if="studioStore.url && studioStore.isActive && !studioStore.isFailed"
+          class="mr-2"
+        ><i-cy-action-record_x16 class="animate-pulse icon-dark-red-500 icon-light-red-500" /></span>
+        <span
+          v-else
+          class="px-2"
+        ><i-cy-object-magic-wand-dark-mode_x16 class="fill-purple-300 stroke-purple-300" /></span>
+        <div class="font-semibold text-base text-gray-800">
+          <span>{{ t('runner.studio.studio').toUpperCase() }}</span>
+          <span class="ml-1"> {{ t('versions.beta').toUpperCase() }}</span>
```
[the remark this answers is not in the document]
Comment (astone123) [reply packages/app/src/runner/studio/StudioControls.vue:15]: Yeah I brought this up to Peter and A. I think the consensus was that we're going to leave it for now

File: packages/app/cypress/e2e/studio/studio.cy.ts [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+      // (2) Click Command
+      cy.get('.command-name-click').should('contain.text', 'click')
+    })
+
+    cy.get('button').contains('Save Commands').click()
+
+    cy.withCtx(async (ctx) => {
+      const spec = await ctx.actions.file.readFileInProject('cypress/e2e/spec.cy.js')
+
+      expect(spec.trim()).to.eq(`
```
Comment (ZachJW34) [inline packages/app/cypress/e2e/studio/studio.cy.ts:33]: Test are working on my Windows machine, but they fail due to the file being read having carriage returns `\r`. A regex replace of `spec.trim().replace(/\r/g, '')` for all instances of the file reads got the tests passing for me.

File: packages/reporter/src/errors/errors.scss (hunk 5 above)
Comment (astone123) [reply packages/reporter/src/errors/errors.scss]: Updated this to only render when the test errors

File: packages/app/cypress/e2e/studio/studio.cy.ts (hunk 13 above)
Comment (astone123) [reply packages/app/cypress/e2e/studio/studio.cy.ts:1]: Looking into this... this test is also failing for me but for a different reason

**Review by ZachJW34 — APPROVED** [review]

**Review by mike-plummer — APPROVED** [review]
Works great, amazing job! 💥

File: packages/app/cypress/e2e/studio/studio.cy.ts (hunk 13 above)
Comment (astone123) [reply packages/app/cypress/e2e/studio/studio.cy.ts:1]: Yeah the tests in this file are really flakey for me... I'm still looking into it. Seems like there's some situations where after `launchStudio` runs, we're not in studio mode. Probably a race condition of sorts

File: packages/app/src/runner/studio/StudioControls.vue [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+            @click="handleRestart"
+          >
+            <i-cy-action-restart_x16 />
+          </button>
+          <template #popper>
+            {{ t('runner.studio.restartStudio') }}
+          </template>
+        </Tooltip>
+
+        <Tooltip
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/app/src/runner/studio/StudioControls.vue:61]: I played around but couldn't find anything weird. Maybe you can demo on a call?

Maybe we want to lift `SelectorPlaygroundTooltip` out of `selector-playground` and make it a more generic `RunnerTooltip` or something? Seems like a nice to have -- not going to block merging this, but let's sync up on this.

Ultimately, I guess this component would go in a design system of sorts.

File: packages/app/cypress/e2e/studio/studio.cy.ts (hunk 13 above)
Comment (lmiller1990) [reply packages/app/cypress/e2e/studio/studio.cy.ts:1]: Couldn't get this exact error. I tried a little commit https://github.com/cypress-io/cypress/pull/23544/commits/24d87183e024857ecd69946854cdc8fc83a69cd9 to reduce flake. These run reliably for me, and on CI :thinking:

Comment (astone123) [reply packages/app/cypress/e2e/studio/studio.cy.ts:1]: Yeah I'm still seeing them flake a bunch in `studio.cy.ts`. I've run them several times and can't get them all to pass at once 🤔 Must just be an issue with my machine. If they're running well in CI then I'd say this isn't a blocker

File: packages/app/src/runner/studio/StudioControls.vue (hunk 18 above)
[the remark this answers is not in the document]
Comment (ZachJW34) [reply packages/app/src/runner/studio/StudioControls.vue:61]: Pretty minor so not blocking but this is what I was seeing.

https://user-images.githubusercontent.com/25158820/187268678-4f6b42ca-a59b-4eda-98b2-c11a7ee23f1b.mov

**Comment by marktnoonan** [comment]
@lmiller1990 Percy shows some changes in general modals that were unexpected for me, [here](https://percy.io/cypress-io/cypress/builds/21076763/changed/1176638386?browser=chrome&browser_ids=23&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=800&widths=400%2C500%2C536%2C800) and [here](https://percy.io/cypress-io/cypress/builds/21076763/changed/1176638146?browser=chrome&browser_ids=23&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=800&widths=400%2C500%2C536%2C800) - I wonder if the change to the `no-help` prop has jostled some things.

**Comment by astone123** [comment]
> @lmiller1990 Percy shows some changes in general modals that were unexpected for me, [here](https://percy.io/cypress-io/cypress/builds/21076763/changed/1176638386?browser=chrome&browser_ids=23&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=800&widths=400%2C500%2C536%2C800) and [here](https://percy.io/cypress-io/cypress/builds/21076763/changed/1176638146?browser=chrome&browser_ids=23&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=800&widths=400%2C500%2C536%2C800) - I wonder if the change to the `no-help` prop has jostled some things.
> 
> In actual usage those components are doing what is expected as far as I can tell, so not a blocker.

Yeah it looks like changing that prop added a few "need help?" links to modals that didn't have them before. For example, the generate spec modal has one that is just a blank link.

File: packages/frontend-shared/src/components/StandardModalHeader.vue [hunk 19]
```diff
@@ -35,10 +37,13 @@ import { useI18n } from '@cy/i18n'
...
 const { t } = useI18n()
 
-defineProps<{
+withDefaults(defineProps<{
   helpLink: string
   helpText: string
-}>()
+  noHelp?: boolean
+}>(), {
+  noHelp: false,
```
Comment (astone123) [inline packages/frontend-shared/src/components/StandardModalHeader.vue:45]: Because this defaults to false, existing modals that don't pass a `helpLink` prop will have links in the modal title that are dead. Thanks @marktnoonan for pointing this out

Comment (marktnoonan) [reply packages/frontend-shared/src/components/StandardModalHeader.vue:45]: Glad you checked! Meanwhile I've edited my comment to remove this line:

> In actual usage those components are doing what is expected as far as I can tell, so not a blocker.

Because I was checking on a different branch.

**Comment by lmiller1990** [comment]
> @lmiller1990 Percy shows some changes in general modals that were unexpected for me, [here](https://percy.io/cypress-io/cypress/builds/21076763/changed/1176638386?browser=chrome&browser_ids=23&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=800&widths=400%2C500%2C536%2C800) and [here](https://percy.io/cypress-io/cypress/builds/21076763/changed/1176638146?browser=chrome&browser_ids=23&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=800&widths=400%2C500%2C536%2C800) - I wonder if the change to the `no-help` prop has jostled some things.

Fixed https://github.com/cypress-io/cypress/pull/23544/commits/06debea6ae740e285082d16ad963a85b6c4740a8
